rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,460 @@
+
+**Spacing values**
+
+- All spacing based properties should accept spacing tokens i.e. `spacing.1`, `spacing.2`, etc.
+- TypeScript auto suggestion hints should also show the `px` value associated with a spacing token, this helps in translating design from Figma to code since designs would have these in `px`.
+- They should also accept `string` values eg. `'1px'`, `'100%'`, `'auto'`, `'1fr'` etc.
+- `number` values should be disallowed eg. `1` as it can lead to confusions _(does this `2` imply `2px` or `spacing.2`)_, if someone wants to use a spacing token it must be prefixed like `spacing.*`.
+- **Tradeoff:** The ability to pass any values here can lead to inconsistencies, where users may be passing values not in tokens (eg. `7px`) or passing the token value directly rather than using the actual token (eg. `2px` instead of `spacing.1`). However, this flexibility is needed in order to ease up building UIs. There are genuine usecases of using percentages and spacings not in spacing tokens (especially in landing pages). Also, spacing tokens are capped (`spacing.11`) so it wouldn't be feasible for users to specify larger spacings (without updating spacing tokens in theme), in such cases they would've to rely on custom components or styling.
+
+**Negative values**
+
+There are some usecases where you may want to pass negative values (eg. negative margin values to create full-bleed layouts that break out of main content). In order to support this we would need spacing values to support negative tokens as well. Instead of negative spacing tokens like `-spacing.1`, `-spacing.2`, etc. we'll let users pass negative values directly eg. `'-4px'`, `'-10%'` etc.
+
+**Few points to note for React (web)**
+
+- We should avoid supporting shorthands to save from confusions _(should I use `mb` or `marginBottom`)_.
+- Flex gap has ~91% global [coverage](https://caniuse.com/?search=flex%20gap) (support added in some browsers mid 2020). Although the support looks decent, this might be problematic for some products supporting these browsers. There is a [post CSS plugin](https://github.com/gavinmcfarland/flex-gap-polyfill) which can polyfill this _(with some known limitations)_. Alternative is to internally translate gap properties to utilize CSS selectors such as [lobotomized owl](https://alistapart.com/article/axiomatic-css-and-lobotomized-owls/) `* + *`.
+- Going forward with native `gap` supports seems like a fine choice here in favor of keeping things simple and native _(keeping it close to CSS `gap` today)_. Tradeoff being, older browsers will be able to render the layout with a degraded experience _(without spacing)_. If users wish to target older browsers they can opt for not using native flex gap and instead use margins (supported with layout styling props).
```

Comment:
How do we plan to let developers know or raise awareness about this? A landing page if created using `gap` may have significant traffic coming from older browsers and those users will see a broken UI if not prevented.

Reply:
Added a suggestion in https://github.com/razorpay/blade/pull/950/commits/7633c183161d7d23854719d1259fff80b7f27617 to add some JSDoc comments to reinforce that we're relying on native flex gap. We can also double clarify this in our layout usage guidelines docs later.